<a href="https://colab.research.google.com/github/Abi-Lowkey/Introduction-to-Machine-Learning-with-Python-Books/blob/main/BAB%201/BAB_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Bab 1: Pengantar Machine Learning**

Implementasi kode

In [1]:
# Import library yang diperlukan
import numpy as np  # Untuk operasi numerik
import pandas as pd  # Untuk manipulasi data
import matplotlib.pyplot as plt  # Untuk visualisasi data
from sklearn.datasets import load_iris  # Untuk memuat dataset Iris
from sklearn.model_selection import train_test_split  # Untuk membagi data menjadi data latih dan data uji
from sklearn.neighbors import KNeighborsClassifier  # Untuk membuat model K-Nearest Neighbors (KNN)


In [2]:
# Memuat dataset Iris menggunakan fungsi load_iris() dari sklearn
iris = load_iris()

# Menampilkan kunci-kunci yang ada dalam dataset Iris
# Dataset Iris disimpan sebagai dictionary, dan kunci-kunci ini menunjukkan elemen-elemen yang tersedia,
# seperti 'data', 'target', 'feature_names', dll.
print("Kunci dataset:", iris.keys())


Kunci dataset: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


In [3]:
# Membagi dataset menjadi data latih (training) dan data uji (testing)
# iris['data'] adalah fitur-fitur (panjang kelopak, lebar kelopak, dll.)
# iris['target'] adalah label (spesies bunga: setosa, versicolor, virginica)

# train_test_split membagi data secara acak:
# - X_train dan y_train digunakan untuk melatih model.
# - X_test dan y_test digunakan untuk menguji performa model.
# Parameter random_state memastikan pembagian data bersifat deterministik (reproducible).
X_train, X_test, y_train, y_test = train_test_split(
    iris['data'], iris['target'], random_state=0
)


In [4]:
# Membuat model k-Nearest Neighbors (k-NN)
# n_neighbors=1 artinya k-NN akan menggunakan 1 tetangga terdekat untuk klasifikasi.
knn = KNeighborsClassifier(n_neighbors=1)

# Melatih model k-NN menggunakan data latih (X_train dan y_train).
# Model akan "mempelajari" hubungan antara fitur dan label dari data latih.
knn.fit(X_train, y_train)


KNeighborsClassifier(n_neighbors=1)

In [5]:
# Melakukan prediksi pada data uji (X_test) menggunakan model k-NN yang telah dilatih
# Hasil prediksi adalah label spesies bunga berdasarkan data uji
predictions = knn.predict(X_test)

# Mengukur akurasi model dengan membandingkan hasil prediksi (predictions) dengan label asli (y_test)
# Metode .score() menghitung proporsi prediksi yang benar
accuracy = knn.score(X_test, y_test)

# Menampilkan hasil prediksi dan akurasi model
print("Prediksi hasil data testing:", predictions)
print("Akurasi data testing:", accuracy)


Prediksi hasil data testing: [2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]
Akurasi data testing: 0.9736842105263158


## **3. Penjelasan Teoritis**
- **Supervised Learning:** Supervised learning merupakan salah satu metode pembelajaran mesin di mana model dilatih menggunakan data yang sudah memiliki label atau pasangan input-output yang diketahui. Dalam metode ini, model mempelajari pola atau hubungan antara data input (fitur) dengan data output (label) untuk kemudian digunakan dalam melakukan prediksi pada data baru yang belum diberi label. Contohnya adalah prediksi spesies bunga berdasarkan karakteristik seperti panjang dan lebar kelopak atau mahkota.

- **Dataset Iris:** Dataset Iris adalah salah satu dataset paling terkenal dalam dunia pembelajaran mesin. Dataset ini terdiri dari 150 sampel bunga dari tiga spesies yang berbeda: Setosa, Versicolor, dan Virginica. Setiap sampel memiliki empat fitur utama, yaitu panjang dan lebar kelopak (sepal) serta panjang dan lebar mahkota (petal). Dataset ini sering digunakan sebagai bahan pembelajaran untuk mengembangkan dan menguji model klasifikasi karena strukturnya yang sederhana namun cukup representatif.

- **k-NN Algorithm:** Algoritma k-Nearest Neighbors (k-NN) adalah salah satu algoritma pembelajaran mesin yang bekerja berdasarkan prinsip kedekatan di ruang fitur. Algoritma ini memprediksi label suatu data berdasarkan label dari k data tetangga terdekatnya dalam ruang fitur. Misalnya, jika k=3, maka algoritma akan melihat tiga tetangga terdekat dari data yang akan diprediksi, dan menentukan label berdasarkan mayoritas label tetangga tersebut. Algoritma k-NN sederhana namun sangat efektif, terutama pada dataset yang tidak terlalu besar dan memiliki pola yang jelas.

4. Insight & Ringkasan
- Proses yang Diperlihatkan: Dalam pembahasan ini, telah dijelaskan langkah-langkah penting dalam membangun model pembelajaran mesin, mulai dari memuat dataset (loading dataset), membagi data menjadi data latih dan data uji (splitting data), melatih model menggunakan algoritma k-Nearest Neighbors (training model), hingga mengevaluasi performa model berdasarkan akurasi data uji (evaluasi model). Semua langkah ini menunjukkan alur kerja yang terstruktur dalam mengembangkan model prediktif.

- Hasil Akurasi: Dengan menggunakan algoritma k-Nearest Neighbors (k-NN) pada dataset Iris, model berhasil mencapai akurasi sebesar 97% pada data uji. Hal ini menunjukkan bahwa k-NN mampu memberikan hasil prediksi yang sangat baik untuk dataset ini.

- Efektivitas Algoritma k-NN: Algoritma k-NN adalah metode yang sederhana namun sangat efektif, terutama saat diterapkan pada dataset kecil seperti Iris. Algoritma ini tidak memerlukan pelatihan yang rumit dan hanya bergantung pada kedekatan data dalam ruang fitur, menjadikannya pilihan yang kuat untuk masalah klasifikasi dengan pola data yang jelas. Meski begitu, untuk dataset yang lebih besar atau memiliki banyak fitur, k-NN mungkin membutuhkan optimisasi tambahan agar tetap efisien.